File: Packages/Features/Reader/Sources/Reader/Screens/Shared/TransportModeSwipe.swift
import CoreGraphics
import Foundation
import SwiftUI

/// Result of a horizontal swipe over the transport control: right hides the seek bar, left brings it back, and
/// anything too short or too vertical leaves the current mode alone.
enum TransportModeSwipeOutcome: Equatable {
    /// Swipe right — drop the expanded seek card for the compact pill.
    case collapse
    /// Swipe left — bring the expanded seek card back.
    case expand
    /// Not a mode swipe; leave the transport as it is.
    case ignore
}

/// Pure geometry for the transport's swipe-to-switch-mode gesture: the commit decision, the rubber band the control
/// rides while the finger is down, and the spring that settles it on release. Kept out of the view so the thresholds
/// are testable — the same split as `PagedReaderNavigation` for the page-turn swipe.
///
/// The feel is modelled on the editing pad's re-dock drag (`EditorChromeView`): the control tracks the finger, leans
/// against a rubber band where there is nothing to switch to, and springs back with the velocity it was released at.
enum TransportModeSwipe {
    /// Horizontal travel that commits a mode change. Roughly the width of one transport button, so the swipe reads as
    /// deliberate but never needs a full sweep across the card.
    static let commitDistance: CGFloat = 44

    /// How far the control can travel toward the mode the swipe would switch to. Well past `commitDistance`, so the
    /// band is barely felt until the swipe has already committed — but finite, so a long sweep can never drag the card
    /// off the screen.
    static let followLimit: CGFloat = 120

    /// How far the control can travel the other way. There is no third mode in that direction, so the control leans a
    /// few points to stay under the finger and no further — the same "this axis isn't a destination" lean the editing
    /// pad has sideways.
    static let leanLimit: CGFloat = 20

    /// Given a finished drag's translation and its fling projection, decide which transport mode the release asks
    /// for.
    ///
    /// Rules:
    /// - A drag that travels further vertically than horizontally is a vertical flick over the transport (or the
    ///   start of a two-finger score gesture), not a mode swipe.
    /// - Commit on `commitDistance` of actual travel; below that, fall back to the predicted end so a fast, short
    ///   flick still switches modes.
    static func outcome(
        translation: CGSize,
        predictedEndTranslation: CGSize,
    ) -> TransportModeSwipeOutcome {
        guard abs(translation.width) >= abs(translation.height) else { return .ignore }

        let travel = abs(translation.width) >= commitDistance
            ? translation.width
            : predictedEndTranslation.width
        guard abs(travel) >= commitDistance else { return .ignore }

        return travel < 0 ? .expand : .collapse
    }

    /// The offset the control renders for a live finger travel of `translation`, given the mode it is currently in.
    ///
    /// Toward the other mode the control is **glued to the finger for the whole commit distance** and only then eases
    /// onto its asymptote. A single `tanh` over the whole range starts resisting almost immediately, and a quick swipe
    /// — which covers a lot of ground fast — left the card visibly trailing the finger: it read as the card catching
    /// on something. Nobody drags past `commitDistance` expecting 1:1 tracking to continue forever, but everybody
    /// expects it up to the point where the swipe means something.
    ///
    /// Away from the other mode there is nothing to reach, so a much stiffer band lets the control lean a few points
    /// and no further.
    static func followOffset(translation: CGFloat, isExpanded: Bool) -> CGFloat {
        // Expanded collapses on a right swipe; compact expands on a left one.
        let towardOtherMode = isExpanded ? translation > 0 : translation < 0
        let sign: CGFloat = translation < 0 ? -1 : 1
        let travel = abs(translation)

        guard towardOtherMode else {
            return sign * leanLimit * CGFloat(tanh(Double(travel / leanLimit)))
        }
        guard travel > commitDistance else { return translation }
        // `tanh` gives the band for free past that point: linear where it takes over from the 1:1 stretch, then
        // flattening onto the asymptote, so the control can never wander further than `followLimit`.
        let give = followLimit - commitDistance
        let beyond = give * CGFloat(tanh(Double((travel - commitDistance) / give)))
        return sign * (commitDistance + beyond)
    }

    /// The settle animation that carries a released control back to zero offset.
    ///
    /// Paced by the same `flickness` as the card's resize, so the two halves of the gesture finish together. They used
    /// to disagree — the offset was timed off its travel distance alone, so a flick left the card sliding home while
    /// the morph was still going (or the reverse), and the mismatch read as the control snagging.
    ///
    /// `bounce: 0` (critically damped) is deliberate: the transport is a slab of controls settling back against the
    /// bottom edge, and overshoot reads as wobble.
    static func releaseAnimation(from offset: CGFloat, releaseVelocity: CGFloat) -> Animation {
        .interpolatingSpring(
            duration: settleDuration(swipeSpeed: releaseVelocity), bounce: 0,
            initialVelocity: settleVelocity(releaseVelocity, travel: -offset),
        )
    }

    /// A spring's initial velocity is expressed in fractions of the REMAINING DISTANCE per second, so raw pt/s has to
    /// be divided by that distance. The dead band keeps a gentle release (which is a release, not a throw) from being
    /// catapulted by what little travel is left.
    ///
    /// A finger still moving *outward* at release yields a velocity pointing away from the settle, and it is KEPT
    /// (clamped to a gentle -1, the editing pad's bound): the control then coasts a touch further before turning
    /// round, which is what momentum looks like. An earlier iteration zeroed it instead — blaming the overshoot for
    /// a release hitch whose real cause was the synchronous score re-layout the preference write used to trigger on
    /// the same frame — and the resulting one-frame stop from full flick speed to zero is precisely what a paused
    /// screen recording shows as the control "catching" at the moment the finger lifts.
    static func settleVelocity(_ velocity: CGFloat, travel: CGFloat) -> CGFloat {
        let deadBand: CGFloat = 80 // pt/s
        guard abs(travel) >= 1, abs(velocity) > deadBand else { return 0 }
        return min(3, max(-1, velocity / travel))
    }

    /// How long the owner should sit on the persisted-preference write after taking a committed swipe. Writing
    /// `showSeekBar` immediately shrinks or grows the room `ReaderRootScreen` reserves for the score
    /// (`bottomControlContentHeight`, 110 ⇄ 44), and in page mode that re-paginates the score — heavy enough to eat
    /// the first frames of the settle spring and the card morph, which read as the control snagging right as the
    /// finger lifts. Longer than the slowest of either release animation (`slowModeSwapDuration`,
    /// `slowSettleDuration`), so the reflow only lands once the transport has visually come to rest; the control
    /// holds the committed mode locally (`previewSeekBar`) until the write catches up, so nothing on screen depends
    /// on it landing promptly.
    static let preferenceCommitDelay = 0.45

    /// How long the acted-out swipe takes to travel out to `commitDistance` before the release spring brings it back
    /// (see `ReaderTransportControl.performHintedModeSwitch`). Short enough to read as one continuous motion with the
    /// morph, long enough that the direction registers.
    static let hintedSwipeOutDuration = 0.16

    /// Slowest swipe that still counts as deliberate, and the speed from which everything is as quick as it gets.
    static let deliberateSwipeSpeed: CGFloat = 200 // pt/s
    static let flickSwipeSpeed: CGFloat = 2000 // pt/s

    /// Card-resize duration at either end of that speed range.
    static let slowModeSwapDuration = 0.35
    static let fastModeSwapDuration = 0.22
    /// Offset-settle duration at either end. Deliberately a touch shorter than the resize at both ends: the offset has
    /// a short distance to cover and looks sluggish if it lingers once the card has stopped changing shape.
    static let slowSettleDuration = 0.30
    static let fastSettleDuration = 0.16

    /// How much of a flick a swipe was: 0 for a deliberate drag, 1 from `flickSwipeSpeed` upward. Direction-blind —
    /// a left swipe is exactly as fast as the right one that mirrors it.
    static func flickness(swipeSpeed: CGFloat) -> Double {
        let speed = min(max(abs(swipeSpeed), deliberateSwipeSpeed), flickSwipeSpeed)
        return Double((speed - deliberateSwipeSpeed) / (flickSwipeSpeed - deliberateSwipeSpeed))
    }

    /// How long the card should take to resize, given how fast the finger was moving when the mode flipped.
    static func modeSwapDuration(swipeSpeed: CGFloat) -> Double {
        paced(slowModeSwapDuration, fastModeSwapDuration, swipeSpeed: swipeSpeed)
    }

    /// How long the released control should take to slide back to zero offset.
    static func settleDuration(swipeSpeed: CGFloat) -> Double {
        paced(slowSettleDuration, fastSettleDuration, swipeSpeed: swipeSpeed)
    }

    private static func paced(_ slow: Double, _ fast: Double, swipeSpeed: CGFloat) -> Double {
        slow + (fast - slow) * flickness(swipeSpeed: swipeSpeed)
    }

    /// The mode swap itself — the card resizing between the seek card and the pill, paced by `swipeSpeed` (the
    /// finger's speed at the moment the mode flipped, or at release for a flick that commits without ever crossing
    /// the threshold), so the resize keeps up with the swipe that asked for it.
    ///
    /// This has to be handed to an explicit `withAnimation` at the point the mode flips, NOT attached as an
    /// `.animation(_:value:)` on the control: the flip happens inside the swipe's `onChanged`, and SwiftUI marks
    /// gesture-driven updates as *continuous* transactions, where implicit animation modifiers are ignored. Attached
    /// implicitly, the card jumped between sizes in a single frame (confirmed frame-by-frame on device) — the mode was
    /// changing correctly, it simply was not being animated.
    static func modeSwapAnimation(swipeSpeed: CGFloat) -> Animation {
        .snappy(duration: modeSwapDuration(swipeSpeed: swipeSpeed), extraBounce: 0)
    }
}

extension View {
    /// Makes a transport row (or the compact pill) swipeable to switch modes.
    ///
    /// `contentShape` extends the swipe to the gaps between the buttons, so the whole row — not just the glyphs —
    /// takes the drag. `highPriorityGesture` is what lets the drag coexist with those buttons: a tap never travels
    /// the gesture's minimum distance, so the button underneath gets it; once the finger does travel, the drag wins
    /// outright and the button it started on is cancelled rather than fired on release (the editing pad's trick).
    func transportModeSwipe(_ gesture: some Gesture) -> some View {
        contentShape(Rectangle())
            .highPriorityGesture(gesture)
    }
}

File: Android/FolinoReaderAndroid/src/main/kotlin/com/keynumber/folino/reader/TransportModeSwipe.kt
package com.keynumber.folino.reader

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.calculateTargetValue
import androidx.compose.animation.core.tween
import androidx.compose.animation.splineBasedDecay
import androidx.compose.foundation.gestures.awaitEachGesture
import androidx.compose.foundation.gestures.awaitFirstDown
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.input.pointer.PointerEventPass
import androidx.compose.ui.input.pointer.changedToUpIgnoreConsumed
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.input.pointer.positionChange
import androidx.compose.ui.input.pointer.util.VelocityTracker
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.unit.Density
import androidx.compose.ui.unit.dp
import com.keynumber.folino.reader.swiftjava.FolinoReaderJNI
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.abs

// The transport's swipe-to-resize gesture: swipe the playback controls right to shrink them to the floating
// cluster, left to bring the seek card back.
//
// EVERY THRESHOLD AND CURVE HERE COMES FROM SWIFT. `ReaderInteractionCore.TransportModeSwipe` owns the commit
// distance, the rubber band, the release velocity and the animation pacing, and iOS's `ReaderTransportControl`
// calls exactly the same functions through `TransportModeSwipe+SwiftUI`. This file is the Compose half — the
// pointer plumbing and the animation objects — and decides nothing about how the gesture feels.

/** `TransportModeSwipeOutcome.rawValue`, as `nativeTransportSwipeOutcome` returns it. */
private const val OUTCOME_COLLAPSE = 0
private const val OUTCOME_EXPAND = 1

/**
 * Live state for one transport. Hoisted to the Reader because the two forms it switches between are two different
 * composables — the committed mode has to decide which of them is mounted, so it cannot live inside either.
 */
class TransportModeSwipeState internal constructor(
    private val scope: CoroutineScope,
    private val density: Density,
) {
    /** The persisted preference. Written by the owner; read here as the mode to fall back on. */
    internal var showSeekBar: Boolean = true

    /** Takes a committed swipe, or declines it (an edit session does). Returns whether it was taken. */
    internal var onSetSeekBarVisible: (Boolean) -> Boolean = { false }

    /**
     * The mode the transport holds from the moment a swipe commits until the preference catches up, and `null` when
     * it is simply following the preference.
     *
     * Local rather than deferred to the preference: that write is delayed on purpose (see
     * `TransportModeSwipe.preferenceCommitDelay`), and a *flicked* swipe — which commits on its projection without
     * ever crossing the threshold — would otherwise have nothing to glide into.
     */
    private var previewSeekBar by mutableStateOf<Boolean?>(null)

    /**
     * Where the control sits. Two sources, never both: the rubber-banded finger travel while a swipe is in progress,
     * and the animation that unwinds it after release. Splitting them keeps the live drag off the animation clock —
     * snapping an `Animatable` once per pointer event would put every frame of the follow behind a coroutine.
     */
    private var isDragging by mutableStateOf(false)
    private var dragOffsetPx by mutableFloatStateOf(0f)
    private val settleOffsetPx = Animatable(0f)

    val offsetPx: Float get() = if (isDragging) dragOffsetPx else settleOffsetPx.value

    private var settleJob: Job? = null

    /** The form to draw right now. */
    val rendersSeekBar: Boolean get() = previewSeekBar ?: showSeekBar

    /**
     * How long the form change caused by the last committed swipe should take — paced by how fast the finger was
     * moving when it let go, so a flick gets a swap that keeps up with it and a deliberate drag gets the long one.
     * Seeded with the deliberate pace, which is also what an acted-out coach mark uses.
     */
    var modeSwapMillis by mutableStateOf(FolinoReaderJNI.nativeTransportModeSwapDurationMillis(0.0))
        private set

    internal fun syncPreference(showSeekBar: Boolean) {
        this.showSeekBar = showSeekBar
        // Drop the local hold once the preference agrees — from here the two say the same thing.
        if (previewSeekBar == showSeekBar) previewSeekBar = null
    }

    /**
     * Switches modes because the swipe coach mark was tapped, by acting the swipe out rather than just flipping:
     * the control slides one commit distance in the direction the finger would have gone, the form changes under it,
     * and the same release animation carries it home. Someone who has never found the gesture sees what it does and
     * where their finger has to go.
     *
     * Paced as a DELIBERATE swipe (speed 0 floors to `deliberateSwipeSpeed`), not a flick: this is a demonstration,
     * and the slow curves are the legible ones.
     */
    fun performHintedModeSwitch() {
        val target = !rendersSeekBar
        if (!onSetSeekBarVisible(target)) return
        // Expanded collapses on a right swipe, compact expands on a left one.
        val travelDp = FolinoReaderJNI.nativeTransportCommitDistance() * if (target) -1 else 1
        val travelPx = with(density) { travelDp.dp.toPx() }
        modeSwapMillis = FolinoReaderJNI.nativeTransportModeSwapDurationMillis(0.0)
        previewSeekBar = target
        settleJob?.cancel()
        isDragging = false
        settleJob = scope.launch {
            settleOffsetPx.animateTo(travelPx, tween(FolinoReaderJNI.nativeTransportHintedSwipeOutMillis()))
            delay(SETTLE_HOLD_MILLIS)
            settleOffsetPx.animateTo(0f, tween(FolinoReaderJNI.nativeTransportSettleDurationMillis(0.0)))
        }
    }

    /**
     * Tracks the finger. The form itself does NOT change until release.
     *
     * iOS flips the card under the finger the moment the swipe passes the commit threshold, because there the
     * transport is ONE view that morphs between two sizes. Android's two forms are two different composables in two
     * different Scaffold slots (`bottomBar` and `floatingActionButton`), so flipping mid-drag would unmount the very
     * node the pointer stream belongs to and the gesture would die half-finished, leaving the control displaced.
     * The rubber band is what says the swipe is going somewhere; the commit lands on release, judged by exactly the
     * same shared `outcome`.
     */
    internal fun onDrag(translationXPx: Float, translationYPx: Float) {
        settleJob?.cancel()
        isDragging = true
        val translationDp = with(density) { translationXPx.toDp().value.toDouble() }
        dragOffsetPx = with(density) {
            FolinoReaderJNI.nativeTransportFollowOffset(translationDp, showSeekBar).dp.toPx()
        }
    }

    /**
     * The node carrying the gesture left the composition (the form changed, the Reader was left) while a swipe was
     * still in progress. There is no release to judge, so the control just goes home rather than staying displaced.
     */
    internal fun onGestureNodeDisposed() {
        if (!isDragging) return
        onDragCancel()
    }

    internal fun onDragEnd(
        translationXPx: Float,
        translationYPx: Float,
        velocityXPx: Float,
        predictedEndXPx: Float,
    ) {
        val translationDp = with(density) { translationXPx.toDp().value.toDouble() }
        val translationYDp = with(density) { translationYPx.toDp().value.toDouble() }
        val predictedDp = with(density) { predictedEndXPx.toDp().value.toDouble() }
        val velocityDp = with(density) { velocityXPx.toDp().value.toDouble() }

        val settled = when (
            FolinoReaderJNI.nativeTransportSwipeOutcome(translationDp, translationYDp, predictedDp)
        ) {
            OUTCOME_COLLAPSE -> if (onSetSeekBarVisible(false)) false else null
            OUTCOME_EXPAND -> if (onSetSeekBarVisible(true)) true else null
            else -> null
        }
        previewSeekBar = settled
        if (settled != null) modeSwapMillis = FolinoReaderJNI.nativeTransportModeSwapDurationMillis(velocityDp)

        // Freeze the damped travel into the settle animation before handing the frame over, so the control does not
        // blink back to its edge as the gesture state evaporates.
        val released = dragOffsetPx
        val releaseVelocity = with(density) {
            FolinoReaderJNI.nativeTransportSettleVelocity(
                velocityDp, -released.toDp().value.toDouble(),
            ).dp.toPx()
        }
        settleJob?.cancel()
        settleJob = scope.launch {
            settleOffsetPx.snapTo(released)
            isDragging = false
            settleOffsetPx.animateTo(
                0f,
                tween(FolinoReaderJNI.nativeTransportSettleDurationMillis(velocityDp)),
                // A finger still moving outward at release keeps its velocity, so the control coasts a touch
                // further before turning round — which is what momentum looks like. The shared clamp is what
                // stops a gentle release from being catapulted by what little travel is left.
                initialVelocity = releaseVelocity,
            )
        }
    }

    internal fun onDragCancel() {
        val released = dragOffsetPx
        settleJob?.cancel()
        settleJob = scope.launch {
            settleOffsetPx.snapTo(released)
            isDragging = false
            settleOffsetPx.animateTo(0f, tween(FolinoReaderJNI.nativeTransportSettleDurationMillis(0.0)))
        }
    }

    private companion object {
        /** A beat at full travel, so the acted-out swipe reads as a gesture rather than a bounce. */
        const val SETTLE_HOLD_MILLIS = 60L
    }
}

/**
 * Holds the swipe state for this Reader, kept in step with the persisted preference.
 *
 * @param showSeekBar the stored preference.
 * @param onSetSeekBarVisible takes a committed swipe (deferring the write — see
 *   `TransportModeSwipe.preferenceCommitDelay`) and answers whether it was taken; declining leaves the control to
 *   animate back, which is what an edit session does.
 */
@Composable
fun rememberTransportModeSwipe(
    showSeekBar: Boolean,
    onSetSeekBarVisible: (Boolean) -> Boolean,
): TransportModeSwipeState {
    val scope = rememberCoroutineScope()
    val density = LocalDensity.current
    val state = remember(scope, density) { TransportModeSwipeState(scope, density) }
    state.syncPreference(showSeekBar)
    state.onSetSeekBarVisible = onSetSeekBarVisible
    return state
}

/**
 * Makes a transport row (or the floating cluster) swipeable to switch modes.
 *
 * The detector runs in the **Initial** pass and claims the gesture only once the travel is horizontal and past touch
 * slop. That is what lets the drag coexist with the buttons underneath: a tap never travels far enough to be claimed,
 * so the button gets it; once the finger does travel, the drag consumes and the button it started on is cancelled
 * rather than fired on release. It is Compose's reading of iOS's `highPriorityGesture` on the same row.
 */
@Composable
fun Modifier.transportModeSwipe(state: TransportModeSwipeState): Modifier {
    DisposableEffect(state) { onDispose { state.onGestureNodeDisposed() } }
    return pointerInput(state) {
        // Android's own fling projection, and the analogue of the `predictedEndTranslation` SwiftUI hands iOS — the
        // shared `outcome` needs one to let a fast, short flick commit without the finger ever travelling the full
        // commit distance.
        val decay = splineBasedDecay<Float>(this)
        awaitEachGesture {
            val down = awaitFirstDown(requireUnconsumed = false, pass = PointerEventPass.Initial)
            val tracker = VelocityTracker()
            tracker.addPosition(down.uptimeMillis, down.position)
            var total = Offset.Zero
            var claimed = false

            while (true) {
                val event = awaitPointerEvent(PointerEventPass.Initial)
                val change = event.changes.firstOrNull { it.id == down.id }
                if (change == null) {
                    // The pointer vanished mid-gesture (another window took the touch, the view was detached).
                    // There is no release to judge, so the control just goes home.
                    if (claimed) state.onDragCancel()
                    break
                }
                if (change.changedToUpIgnoreConsumed()) {
                    if (claimed) {
                        val velocityX = tracker.calculateVelocity().x
                        state.onDragEnd(
                            translationXPx = total.x,
                            translationYPx = total.y,
                            velocityXPx = velocityX,
                            predictedEndXPx = total.x + decay.calculateTargetValue(0f, velocityX),
                        )
                    }
                    break
                }
                total += change.positionChange()
                tracker.addPosition(change.uptimeMillis, change.position)
                if (!claimed && abs(total.x) > viewConfiguration.touchSlop && abs(total.x) >= abs(total.y)) {
                    claimed = true
                }
                if (claimed) {
                    change.consume()
                    state.onDrag(total.x, total.y)
                }
            }
        }
    }
}
